### PR TITLE
JP-3657: Part 1 of a NIRSpec BOTS speedup

### PR DESCRIPTION
**Pull request by t-brandt**
Partially resolves [JP-3657](https://jira.stsci.edu/browse/JP-3657)

Partially addresses #8559

This PR enables the reuse of aperture correction objects for spectroscopic data with multiple integrations, specifically targeted at BOTS data.  Most of the time in a BOTS reduction to x1d files goes into constructing multidimensional interpolating splines for every wavelength and for every integration.  Reusing these across integrations saves a factor of ~10 in BOTS runtime.  A further factor of maybe 2 or 3, larger if a background is fit, will come from a rewrite of extract1d (hence this PR only partially addresses the JIRA ticket).

**Checklist for PR authors (skip items if you don't have permissions or they are not applicable)**
- [x] added entry in `CHANGES.rst` within the relevant release section
- [x] updated or added relevant tests
- [x] updated relevant documentation
- [x] added relevant milestone
- [x] added relevant label(s)
- [x] ran regression tests, post a link to the Jenkins job below.
      [How to run regression tests on a PR](https://github.com/spacetelescope/jwst/wiki/Running-Regression-Tests-Against-PR-Branches)
- [x] All comments are resolved
- [ ] Make sure the JIRA ticket is [resolved properly](https://github.com/spacetelescope/jwst/wiki/How-to-resolve-JIRA-issues)

### CODE REVIEW

File: jwst/extract_1d/extract.py [hunk 1]
```diff
@@ -3736,6 +3736,13 @@ def create_extraction(extract_ref_dict,
         use_source_posn = False
         log.info(f"Setting use_source_posn to False for source type {source_type}")
 
+    # Turn off use_source_posn if working on non-primary NRS fixed slits
```
Comment (t-brandt) [inline jwst/extract_1d/extract.py:3739]: I am not sure why these five lines are showing up.  I did not make these edits, and I freshly pulled the jwst master branch just before committing and submitting the PR.

Comment (melanieclarke) [reply jwst/extract_1d/extract.py:3739]: It's probably a merging issue.  Those lines were removed in #8467, which touched many parts of the pipeline code.

File: jwst/extract_1d/extract.py [hunk 2]
```diff
@@ -3896,27 +3898,57 @@ def create_extraction(extract_ref_dict,
...
+                    log.info("Tabulating aperture correction for use in multiple integrations.")
+                except AttributeError:
+                    log.info("Computing aperture correction.")
+                    apcorr.apply(spec.spec_table)
+
+        # Save previous ra, dec, wavelength in case we can reuse
+        # the aperture correction object.
+        ra_last = ra
+        dec_last = dec
+        wl_last = wl
```
[the remark this answers is not in the document]
Comment (melanieclarke) [reply jwst/extract_1d/extract.py:3951]: This is also why the unit tests are failing.

File: jwst/extract_1d/apply_apcorr.py [hunk 3]
```diff
@@ -227,6 +228,51 @@ def _approx_func(wavelength: float, size: float, pixel_phase: float) -> RectBiva
...
+
+        Parameters
+        ----------
+        spec_table : `~fits.FITS_rec`
+            Table of aperture corrections values from apcorr reference file.
+
+        """
+
+        if self.tabulated_correction is None:
+            raise ValueError("Cannot call apply_tabulated_correction without first calling tabulate_correction")
```
[the remark this answers is not in the document]
Comment (t-brandt) [reply jwst/extract_1d/apply_apcorr.py]: I implemented this but added the closing quotes on the top line

[the remark this answers is not in the document]
Comment (melanieclarke) [reply jwst/extract_1d/apply_apcorr.py]: Good catch! Sorry!

**Comment by t-brandt** [comment]
I refactored the code slightly so that apcorr.apply() takes use_tabulated as an (optional) argument.  Please tell me if you think this approach is better.  If so, we can double-check the test results.

I think we should be able to write a unit test to verify that calling tabulate_correction and .apply(use_tabulated=True) gives the same results as .apply(usetabulated=False).  It would be best to do this on a data set with a nontrivial aperture correction.

**Comment by melanieclarke** [comment]
> I refactored the code slightly so that apcorr.apply() takes use_tabulated as an (optional) argument. Please tell me if you think this approach is better. If so, we can double-check the test results.

These changes look sensible to me - thanks!

I'll be out next week, but can check back in after if there is still testing work that needs to be done.

**Comment by t-brandt** [comment]
Adding a unit test to verify that the tabulated and non-tabulated corrections are the same is harder than I thought.  This will require a new dummy data set (since none of the ones there currently have the keys in the data tables).  I was hoping it would just be a couple of lines.  I could generate a data set and add it to the repository, and then write a test to verify that the answer is the same in the various ways of computing an aperture correction.  Should I do this?

**Comment by melanieclarke** [comment]
> Adding a unit test to verify that the tabulated and non-tabulated corrections are the same is harder than I thought. This will require a new dummy data set (since none of the ones there currently have the keys in the data tables). I was hoping it would just be a couple of lines. I could generate a data set and add it to the repository, and then write a test to verify that the answer is the same in the various ways of computing an aperture correction. Should I do this?

I'm not very familiar with the current apcorr tests. Is it possible to just add the keys you need to the existing mock data sets?

**Comment by t-brandt** [comment]
The current tests simply check that the interpolating aperture correction function is unity at a random wavelength (i.e. that there is no correction).  A proper test, to my mind, for the current change is that the aperture correction applied to a grid of wavelengths is the same whether applied in the old or in the new way.  The function that applies the correction has a variety of fields hard-coded in in the form of a FITS table that it assumes as input; such a table is never actually called or referenced in the current checks.  So I think it would be a bit of a pain to implement and would require a new dummy reference file.  If it is important to do this I can give it a go.

**Comment by drlaw1558** [comment]
It may be simpler then to cover this in the regression test framework (which should happen automatically) rather than as a unit test.

**Comment by melanieclarke** [comment]
I see, the current unit tests don't cover any of this functionality yet. I agree it would take some development time to work up appropriate mock test data.  I think it would be nice to have, but it's not critical for this change.

**Comment by melanieclarke** [comment]
I think all comments have been addressed for this PR.  I'll regression tests when the system is free.

**Review by melanieclarke — APPROVED** [review]
No failures at all in the regression tests!  I'm not sure I've ever seen that before.

LGTM!

**Pull request merged**